openshift/console
# HPA In-Context (Delete | Edit | Topology Sidebar)

Combined for easy reviewing. Former PRs: #6037, #6039, #6087

**Fixes**: 
https://issues.redhat.com/browse/ODC-4270
https://issues.redhat.com/browse/ODC-4271
https://issues.redhat.com/browse/ODC-4272

**Analysis / Root cause**: 

- As a user, I want a way to delete an in-context HPA.
- As a user, I'd like to be able to in-context edit an associated HPA.
- As a user, I'd like to be able to see my HPA resource in the sidebar and not be able to change pods as the HPA will fight me

**Solution Description**: 

- Add a Kebab action and modal to support deleting the associated HPA on a workload.
- Using the same form from the Add HPA, allow editing.
- Update to add a resource section for HPAs and to stop manual scaling of the pods; through kebab and scale arrows on the pod donut that is on the details tab.

**Screen shots / Gifs for design review**: 
@openshift/team-devconsole-ux & @lwrigh 

![Screen Shot 2020-07-22 at 7 32 11 AM](https://user-images.githubusercontent.com/8126518/88171545-8195fe80-cbed-11ea-8391-0186013f3fa8.png)
![Screen Shot 2020-07-22 at 7 32 19 AM](https://user-images.githubusercontent.com/8126518/88171547-82c72b80-cbed-11ea-87d0-ee81969c952b.png)

When a custom metric is in play, we auto select the YAML - they can still use the Form when they so desire to get access to cpu / memory (if applicable), the hpa name and min/max replica counts.
![Screen Shot 2020-07-22 at 7 50 51 AM](https://user-images.githubusercontent.com/8126518/88173176-244f7c80-cbf0-11ea-9b5c-b6183bbe18a7.png)

Sidebar:
![Screen Shot 2020-07-23 at 1 24 52 PM](https://user-images.githubusercontent.com/8126518/88319858-c2277200-ccea-11ea-96fc-4a071f539c2f.png)
![Screen Shot 2020-07-23 at 1 25 06 PM](https://user-images.githubusercontent.com/8126518/88319864-c3589f00-ccea-11ea-9a3a-027e8db6a215.png)
![Screen Shot 2020-07-23 at 1 42 05 PM](https://user-images.githubusercontent.com/8126518/88319867-c3f13580-ccea-11ea-9d1a-a441d5899863.png)

**Unit test coverage report**: 

```
  doesHpaMatch checks if it aligns to a workload
    ✓ expect not to match when hpa does not target workload
    ✓ expect to match when hpa does target workload
  hasCustomMetrics accurately determines if an hpa contains non-default metrics
    ✓ expect no metrics to mean no custom metrics
    ✓ expect cpu and memory metrics to not be custom
    ✓ expect any unknown metric to be custom
    ✓ expect an unknown metric to trump known metrics

------------------------------------------------------------|----------|----------|----------|----------|-------------------|
File                                                        |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------------------------------------|----------|----------|----------|----------|-------------------|
  src/components/hpa/hpa-utils.ts                           |    97.26 |    90.63 |      100 |    96.83 |            92,100 |
------------------------------------------------------------|----------|----------|----------|----------|-------------------|
```

**Test setup:**

- Add an HPA to a D/DC Workload that is in Topology
- Use the Kebab action to delete || edit
- See Topology Sidebar Details page for no pod scaling and new text

**Browser conformance**:
